### PR TITLE
The `HUB` debugging never stops

### PR DESCRIPTION
**Pull request by OlivierBBB**

### CODE REVIEW

File: hub/constraints/instruction-handling/call/finishing_touches/abort.lisp [hunk 1]
```diff
@@ -28,13 +28,13 @@
                     (vanishes!    (shift    account/ROMLEX_FLAG      CALL_2nd_callee_account_row___abort_will_revert___row_offset))
                     (vanishes!    (shift    account/TRM_FLAG         CALL_2nd_callee_account_row___abort_will_revert___row_offset))
                     (vanishes!    (shift    account/RLPADDR_FLAG     CALL_2nd_callee_account_row___abort_will_revert___row_offset))
-                    (DOM-SUB-stamps---standard                       CALL_2nd_callee_account_row___abort_will_revert___row_offset
+                    (DOM-SUB-stamps---revert-with-current            CALL_2nd_callee_account_row___abort_will_revert___row_offset
```
Comment (OlivierBBB) [inline hub/constraints/instruction-handling/call/finishing_touches/abort.lisp:31]: This was a real mistake. It's a revert `DOM/SUB` stamp we require

File: hub/constraints/instruction-handling/call/generalities/1st_set_of_account_rows.lisp [hunk 2]
```diff
@@ -42,8 +42,8 @@
                     (if-not-zero    (scenario-shorthand---CALL---balance-update-not-required)
                                     (account-same-balance            CALL_1st_caller_account_row___row_offset))
                     (if-not-zero    (scenario-shorthand---CALL---balance-update-required)
-                                    (account-decrement-balance-by    CALL_1st_caller_account_row___row_offset    (call-instruction---STACK-value-lo)))
-                    ))
+                                    (account-decrement-balance-by    CALL_1st_caller_account_row___row_offset
+                                                                     (* (call-instruction---is-CALL) (call-instruction---STACK-value-lo))))))
```
Comment (OlivierBBB) [inline hub/constraints/instruction-handling/call/generalities/1st_set_of_account_rows.lisp:46]: This fixes the bug where `CALLCODE` produced a value transfer (_caller_ side)

File: hub/constraints/instruction-handling/call/generalities/1st_set_of_account_rows.lisp [hunk 3]
```diff
@@ -74,7 +74,8 @@
                     (if-not-zero    (scenario-shorthand---CALL---balance-update-not-required)
                                     (account-same-balance            CALL_1st_callee_account_row___row_offset))
                     (if-not-zero    (scenario-shorthand---CALL---balance-update-required)
-                                    (account-increment-balance-by    CALL_1st_callee_account_row___row_offset    (call-instruction---STACK-value-lo)))))
+                                    (account-increment-balance-by    CALL_1st_callee_account_row___row_offset
+                                                                     (* (call-instruction---is-CALL) (call-instruction---STACK-value-lo))))))
```
Comment (OlivierBBB) [inline hub/constraints/instruction-handling/call/generalities/1st_set_of_account_rows.lisp:78]: This fixes the bug where `CALLCODE` produced a value transfer (_callee_ side)

File: hub/constraints/instruction-handling/call/generalities/3rd_set_of_account_rows.lisp [hunk 4]
```diff
@@ -18,7 +18,8 @@
 (defun    (call-instruction---summon-callee-account-thrice)    (*    PEEK_AT_SCENARIO
                                                                      scenario/CALL_SMC_FAILURE_CALLER_WILL_REVERT))
 
-(defconstraint    call-instruction---3rd-callee-account-operation                     (:guard (call-instruction---summon-both-account-rows-twice-or-more))
+(defconstraint    call-instruction---3rd-callee-account-operation
+                  (:guard (call-instruction---summon-callee-account-thrice))
```
Comment (OlivierBBB) [inline hub/constraints/instruction-handling/call/generalities/3rd_set_of_account_rows.lisp:22]: This was using the wrong guard, thus triggering those constraints in other cases besides `SMC_FAILURE_WILL_REVERT`.

File: hub/constraints/instruction-handling/call/generalities/universal.lisp [hunk 5]
```diff
@@ -103,15 +103,15 @@
...
+                                  (set-STP-instruction-call   CALL_misc_row___row_offset                                           ;; relative row offset
+                                                              (call-instruction---STACK-instruction)                               ;; instruction
+                                                              (call-instruction---STACK-gas-hi)                                    ;; max gas allowance argument, high part
+                                                              (call-instruction---STACK-gas-lo)                                    ;; max gas allowance argument, low  part
+                                                              (call-instruction---STACK-value-hi)                                  ;; value to transfer, high part
+                                                              (call-instruction---STACK-value-lo)                                  ;; value to transfer, low  part
+                                                              (* (call-instruction---callee-exists) (call-instruction---is-CALL))  ;; bit indicating target account existence
```
Comment (OlivierBBB) [inline hub/constraints/instruction-handling/call/generalities/universal.lisp:112]: HUB side of the changes in the HUB/STP interface where EXISTS for the target of a CALL-type instruction only matters if the instruction is precisely `CALL`

File: hub/constraints/instruction-handling/call/specialized/missing-context-parameters.lisp [hunk 6]
```diff
@@ -46,7 +46,7 @@
                                                                       (*    (call-instruction---is-DELEGATECALL)    (call-instruction---current-caller-address-lo))
                                                                       (*    (call-instruction---is-STATICCALL)      (call-instruction---current-address-lo))))
 
-(defun    (call-instruction---new-call-value)                   (+    (*    (call-instruction---is-CALL)            (call-instruction---STACK-value-lo)
+(defun    (call-instruction---new-call-value)                   (+    (*    (call-instruction---is-CALL)            (call-instruction---STACK-value-lo))
                                                                       (*    (call-instruction---is-CALLCODE)        (call-instruction---STACK-value-lo))
                                                                       (*    (call-instruction---is-DELEGATECALL)    (call-instruction---current-call-value))
-                                                                      (*    (call-instruction---is-STATICCALL)      0))))
+                                                                      (*    (call-instruction---is-STATICCALL)      0)))
```
Comment (OlivierBBB) [inline hub/constraints/instruction-handling/call/specialized/missing-context-parameters.lisp:52]: Parenthesis mistake 😭

File: hub/constraints/instruction-handling/copy/codecopy.lisp [hunk 7]
```diff
@@ -45,6 +45,6 @@
                   (begin
                     (eq!        (shift account/DEPLOYMENT_NUMBER      ROFF_CODECOPY_NO_XAHOY_ACCOUNT_ROW) (shift context/BYTE_CODE_DEPLOYMENT_NUMBER     ROFF_CODECOPY_NO_XAHOY_CONTEXT_ROW))
                     (eq!        (shift account/DEPLOYMENT_STATUS      ROFF_CODECOPY_NO_XAHOY_ACCOUNT_ROW) (shift context/BYTE_CODE_DEPLOYMENT_STATUS     ROFF_CODECOPY_NO_XAHOY_CONTEXT_ROW))
-                    (eq!        (shift account/DEPLOYMENT_STATUS      ROFF_CODECOPY_NO_XAHOY_ACCOUNT_ROW) (shift context/BYTE_CODE_CODE_FRAGMENT_INDEX   ROFF_CODECOPY_NO_XAHOY_CONTEXT_ROW))
+                    (eq!        (shift account/CODE_FRAGMENT_INDEX    ROFF_CODECOPY_NO_XAHOY_ACCOUNT_ROW) (shift context/BYTE_CODE_CODE_FRAGMENT_INDEX   ROFF_CODECOPY_NO_XAHOY_CONTEXT_ROW))
```
Comment (OlivierBBB) [inline hub/constraints/instruction-handling/copy/codecopy.lisp:48]: Plain old copy paste error

File: hub/constraints/instruction-handling/keccak.lisp [hunk 8]
```diff
@@ -18,21 +18,28 @@
...
 (defun   (keccak-instruction---trigger_MMU)           (* (- 1 XAHOY) (keccak-instruction---mxp-MTNTOP)))
 (defun   (keccak-instruction---no-stack-exceptions)   (* PEEK_AT_STACK stack/KEC_FLAG (- 1 stack/SUX stack/SOX)))
 
 (defconstraint    keccak-instruction---setting-stack-pattern (:guard (keccak-instruction---no-stack-exceptions))
                   (stack-pattern-2-1))
 
+(defconstraint    keccak-instruction---allowable-exceptions (:guard (keccak-instruction---no-stack-exceptions))
+                  (eq!    XAHOY
+                          (+    stack/MXPX
+                                stack/OOGX)))
```
Comment (OlivierBBB) [inline hub/constraints/instruction-handling/keccak.lisp:41]: Make the exceptions explicit.

File: stp/constraints.lisp [hunk 9]
```diff
@@ -105,36 +105,35 @@
...
+
+(defconstraint    CALL-type---debug---non-value-transferring-opcodes-have-zero-value    ()
+                  (if-not-zero    (+    IS_DELEGATECALL    IS_STATICCALL)
+                                  (begin
+                                    (vanishes!    VAL_HI)
+                                    (vanishes!    VAL_LO))))
+
+(defconstraint    CALL-type---debug---account-existence-only-matters-for-CALL    ()
+                  (if-not-zero    (+    IS_CALLCODE    IS_DELEGATECALL    IS_STATICCALL)
+                                  (vanishes!    EXISTS)))
```
Comment (OlivierBBB) [inline stp/constraints.lisp:122]: Non `debug` "debug" constraints.

File: stp/constraints.lisp [hunk 10]
```diff
@@ -205,7 +204,8 @@
                                                                  GAS_CONST_G_COLD_ACCOUNT_ACCESS))
 (defun (call---gas-value-transfer-cost)          (*    (call---transfers-value)
                                                        GAS_CONST_G_CALL_VALUE))
-(defun (call---gas-new-account-cost)             (*    (-    1    EXISTS)
+(defun (call---gas-new-account-cost)             (*    IS_CALL
+                                                       (-    1    EXISTS)
```
Comment (OlivierBBB) [inline stp/constraints.lisp:208]: Among all CALL-type instructions only `CALL` may lead to an account being created. This fixes a bug where `CALLCODE` would sometimes cost an extra `25_000` gas despite no such surcharge applying.

File: hub/constraints/instruction-handling/call/finishing_touches/abort.lisp (hunk 1 above)
Comment (letypequividelespoubelles) [reply hub/constraints/instruction-handling/call/finishing_touches/abort.lisp:31]: indeed

File: hub/constraints/instruction-handling/call/specialized/missing-context-parameters.lisp (hunk 6 above)
Comment (letypequividelespoubelles) [reply hub/constraints/instruction-handling/call/specialized/missing-context-parameters.lisp:52]: The worst to debug :)

File: stp/constraints.lisp [hunk 11]
```diff
@@ -46,7 +46,7 @@
 ;;    2.3 heartbeat    ;;
 ;;                     ;;
 ;;;;;;;;;;;;;;;;;;;;;;;;;
-(defconstraint first-row (:domain {0})
+(defconstraint first-row (:domain {0}) ;; ""
```
Comment (letypequividelespoubelles) [inline stp/constraints.lisp:49]: What is the "" ?

File: stp/constraints.lisp (hunk 9 above)
Comment (letypequividelespoubelles) [reply stp/constraints.lisp:122]: ? What do you mean. Why they have debug is their names if it's not a debug constraint ?

**Review by letypequividelespoubelles — APPROVED** [review]
only minor comments

File: stp/constraints.lisp (hunk 9 above)
Comment (OlivierBBB) [reply stp/constraints.lisp:122]: I've done that in a few places ... having constraints with `---debug---` in their name. Typically for stuff that maybe shouldn't be debug after all.

File: stp/constraints.lisp (hunk 11 above)
Comment (OlivierBBB) [reply stp/constraints.lisp:49]: Never mind them, I sometimes add them to help my IDE with the colors ...

**Pull request merged**